aten/src/ATen/native/mps/operations/UnaryOps.mm

```diff
@@ -406,4 +407,48 @@ Tensor logit_mps(const Tensor& self, c10::optional<double> eps) {
                 });
 }
 
+TORCH_IMPL_FUNC(cumprod_out_mps)
+(const Tensor& self, int64_t dim, c10::optional<ScalarType> dtype, const Tensor& result) {
+  bool macOS13_3_plus = is_macos_13_or_newer(MacOSVersion::MACOS_VER_13_3_PLUS);
+  auto nDims = self.dim();
+  auto wrapped_dim = maybe_wrap_dim(dim, nDims);
+  TORCH_CHECK(wrapped_dim >= 0 && wrapped_dim < std::max(1LL, self.ndimension()),
+              "Expected wrapped dim to be between 0 and ",
+              self.ndimension(),
+              " but got ",
+              wrapped_dim,
+              "(original dim is ",
+              dim,
+              ")");
+  if (!is_macos_13_or_newer()) {
+    TORCH_WARN_ONCE("torch.cumprod supported by MPS on MacOS 13+, please upgrade");
+    auto cpu_result = self.to(at::Device(kCPU)).cumprod(dim, dtype);
+    at::_copy_from_and_resize(cpu_result, result);
+    return;
+  }
+  auto input = dtype.has_value() ? self.to(dtype.value()) : self;
+
+  // issue #103810551: cumprod is horribly broken for int8, int16 and as chances for overflow is pretty high, cast to
```

Comment:
I'm not sure what `issue #103810551` is and where I want to check if this applies to cumprod as well as cumsum. So far I assumed it did but seeing the actual issue in whatever tracker it lies would help

Reply:
from the work done on https://github.com/kulinseth/pytorch/pull/419, it appears constraint does hold for both functions so I will merge them into a single one as is done in this pr